File: Classes/EVEOnlineAPI/Character/EVESkillQueue.m
//
//  EVESkillQueue.m
//  EVEOnlineAPI
//
//  Created by Artem Shimanski on 7/24/10.
//  Copyright 2010 __MyCompanyName__. All rights reserved.
//

#import "EVESkillQueue.h"


@implementation EVESkillQueueItem

+ (NSDictionary*) scheme {
	static NSDictionary* scheme = nil;
	if (!scheme)
		scheme = @{@"queuePosition":@{@"type":@(EVEXMLSchemePropertyTypeScalar)},
				   @"typeID":@{@"type":@(EVEXMLSchemePropertyTypeScalar)},
				   @"level":@{@"type":@(EVEXMLSchemePropertyTypeScalar)},
				   @"startSP":@{@"type":@(EVEXMLSchemePropertyTypeScalar)},
				   @"endSP":@{@"type":@(EVEXMLSchemePropertyTypeScalar)},
				   @"startTime":@{@"type":@(EVEXMLSchemePropertyTypeDate)},
				   @"endTime":@{@"type":@(EVEXMLSchemePropertyTypeDate)}};
	
	return scheme;
}

@end


@implementation EVESkillQueue

+ (NSDictionary*) scheme {
	static NSDictionary* scheme = nil;
	if (!scheme)
		scheme = @{@"skillQueue":@{@"type":@(EVEXMLSchemePropertyTypeRowset), @"class":[EVESkillQueueItem class], @"elementName": @"skillqueue"}};
	
	return scheme;
}

- (NSTimeInterval) timeLeft {
	NSDate *endTime = [[self.skillQueue lastObject] endTime];
	return [endTime timeIntervalSinceDate:[self.eveapi serverTimeWithLocalTime:[NSDate date]]];
}


@end

File: EVEAPI/EVEAPI/EVESkillQueue.swift
//
//  EVESkillQueue.swift
//  EVEAPI
//
//  Created by Artem Shimanski on 29.11.16.
//  Copyright © 2016 Artem Shimanski. All rights reserved.
//

import UIKit

public class EVESkillQueueItem: EVEObject {
	public var queuePosition: Int = 0
	public var typeID: Int = 0
	public var level: Int = 0
	public var startSP: Int = 0
	public var endSP: Int = 0
	public var startTime: Date = Date.distantPast
	public var endTime: Date = Date.distantPast
	
	public required init?(dictionary:[String:Any]) {
		super.init(dictionary: dictionary)
	}
	
	public required init?(coder aDecoder: NSCoder) {
		super.init(coder: aDecoder)
	}
	
	override public func scheme() -> [String:EVESchemeElementType] {
		return [
		"queuePosition":EVESchemeElementType.Int(elementName:nil, transformer:nil),
		"typeID":EVESchemeElementType.Int(elementName:nil, transformer:nil),
		"level":EVESchemeElementType.Int(elementName:nil, transformer:nil),
		"startSP":EVESchemeElementType.Int(elementName:nil, transformer:nil),
		"endSP":EVESchemeElementType.Int(elementName:nil, transformer:nil),
		"startTime":EVESchemeElementType.Date(elementName:nil, transformer:nil),
		"endTime":EVESchemeElementType.Date(elementName:nil, transformer:nil),
		]
	}
}


public class EVESkillQueue: EVEResult {
	public var skillQueue: [EVESkillQueueItem] = []
	public var timeLeft: TimeInterval = 0

	public required init?(dictionary:[String:Any]) {
		super.init(dictionary: dictionary)
	}
	
	public required init?(coder aDecoder: NSCoder) {
		super.init(coder: aDecoder)
	}
	
	override public func scheme() -> [String:EVESchemeElementType] {
		return [
			"skillQueue":EVESchemeElementType.Rowset(elementName: nil, type: EVESkillQueueItem.self, transformer: nil),
			"timeLeft":EVESchemeElementType.Double(elementName:nil, transformer:nil),
		]
	}
}

